shiny/render/_dataframe.py

```diff
@@ -258,29 +402,334 @@ class data_frame(Renderer[DataFrameResult]):
+        @reactive.calc
+        def self_data() -> pd.DataFrame:
+            value = self._value()
+            req(value)
+
+            if not isinstance(value, (DataGrid, DataTable)):
+                raise TypeError(
+                    f"Unsupported type returned from render function: {type(value)}. Expected `DataGrid` or `DataTable`"
+                )
+
+            if not isinstance(value.data, pd.DataFrame):
+                raise TypeError(f"Unexpected type for self._data: {type(value.data)}")
+
+            return value.data
+
+        self.data = self_data
+
+        @reactive.calc
+        def self_table_mode() -> DataFrameMode:
+            value = self._value()
+            req(value)
+            if not isinstance(value, (DataGrid, DataTable)):
+                raise TypeError(
+                    f"Unsupported type returned from render function: {type(value)}. Expected `DataGrid` or `DataTable`"
+                )
+
+            return value.mode
+
+        self.table_mode = self_table_mode
+
+        @reactive.calc
+        def self_input_selected() -> SelectedIndicies | None:
+            mode = self.table_mode()
+            if mode == "none":
+                return None
+            return {
+                "rows": self._get_session().input[f"{self.output_id}_selected_rows"](),
+                "columns": None,
+            }
+
+        self.input_selected = self_input_selected
+
+        @reactive.calc
+        def self_data_selected() -> pd.DataFrame:
+            indicies = self.input_selected()
+            if indicies is None:
+                req(False)
+                raise RuntimeError("This should never be reached for typing purposes")
+
+            data_selected = self.data_patched()
+            if indicies["rows"] is not None:
+                data_selected = data_selected.iloc[list(indicies["rows"])]
+            if indicies["columns"] is not None:
+                data_selected = data_selected.iloc[:, list(indicies["columns"])]
+            return data_selected
+
+        self.data_selected = self_data_selected
+
+        @reactive.calc
+        def self_data_patched() -> pd.DataFrame:
+            # Enable copy-on-write mode for the data;
+            # Use `deep=False` to avoid copying the full data; CoW will copy the necessary data when modified
+            with pd.option_context("mode.copy_on_write", True):
+                data = self.data().copy(deep=False)
+                for cell_patch in self.cell_patches():
+                    data.iat[  # pyright: ignore[reportUnknownMemberType]
+                        cell_patch.row_index,
+                        cell_patch.column_index,
+                    ] = cell_patch.value
+
+                return data
+
+        self.data_patched = self_data_patched
+
+    def _get_session(self) -> Session:
+        if self._session is None:
+            raise RuntimeError(
+                "The data frame's session can only be accessed within a reactive context"
```

Comment:
I think this error message is saying something different from the actual issue -- the data frame renderer must be initialized in a reactive context / session. (And perhaps it should throw in the init method instead?)